From: Patrick Brunschwig <patrick@enigmail.net>
Date: Tue, 13 Feb 2018 11:59:58 +0100
Subject: Bug 745: - fixed matching algorithm for getting secret keys - click
 on "Open Key Properties" doesn't work

(cherry picked from commit 9b958e59efb6b0f0dd3ca394952c459ed8c8e839)
---
 package/keyRing.jsm                     | 15 +++++++++++++++
 package/keyUsability.jsm                | 32 +++++++++++++-------------------
 package/windows.jsm                     |  4 ++++
 ui/content/enigmailMsgComposeOverlay.js |  2 +-
 4 files changed, 33 insertions(+), 20 deletions(-)

diff --git a/package/keyRing.jsm b/package/keyRing.jsm
index 9329d64..8af14aa 100644
--- a/package/keyRing.jsm
+++ b/package/keyRing.jsm
@@ -294,6 +294,21 @@ var EnigmailKeyRing = {
     return res;
   },
 
+  /**
+   * Specialized function for getSecretKeyByUserId() that takes into account
+   * the specifics of email addresses in UIDs.
+   *
+   * @param emailAddr: String - email address to search for without any angulars
+   *                            or names
+   *
+   * @return KeyObject with the found key, or null if no key found
+   */
+  getSecretKeyByEmail: function(emailAddr) {
+    let searchTerm = "(<" + emailAddr + ">| " + emailAddr + "$|^" + emailAddr + "$)";
+
+    return this.getSecretKeyByUserId(searchTerm);
+  },
+
   /**
    * get the "best" possible secret key for a given user ID
    *
diff --git a/package/keyUsability.jsm b/package/keyUsability.jsm
index 05f83cb..a9dfda9 100644
--- a/package/keyUsability.jsm
+++ b/package/keyUsability.jsm
@@ -45,27 +45,24 @@ var EnigmailKeyUsability = {
     if (!enigmailSvc) return [];
 
     let result = keySpecArr.reduce(function(p, keySpec) {
-      let keys;
+      let key;
 
       if (keySpec.search(/^(0x)?[0-9A-F]{8,40}$/i) === 0) {
         let key = getKeyRing().getKeyById(keySpec);
         if (!key) return p;
-        keys = [key];
       }
       else {
-        keys = getKeyRing().getKeysByUserId(keySpec);
-        if (keys.length === 0) return p;
+        key = getKeyRing().getSecretKeyByEmail(keySpec);
+        if (!key) return p;
       }
 
       let maxExpiry = Number.MIN_VALUE;
       let maxKey = null;
 
-      for (let i in keys) {
-        let ex = keys[i].getKeyExpiry();
-        if (ex > maxExpiry) {
-          maxExpiry = ex;
-          maxKey = keys[i];
-        }
+      let ex = key.getKeyExpiry();
+      if (ex > maxExpiry) {
+        maxExpiry = ex;
+        maxKey = key;
       }
 
       if (maxExpiry < now + (DAY * numDays) && maxExpiry >= now) p.push(maxKey);
@@ -203,22 +200,19 @@ var EnigmailKeyUsability = {
     if (!enigmailSvc) return [];
 
     let result = keySpecArr.reduce(function(p, keySpec) {
-      let keys;
+      let key;
 
       if (keySpec.search(/^(0x)?[0-9A-F]{8,40}$/i) === 0) {
-        let key = getKeyRing().getKeyById(keySpec);
+        key = getKeyRing().getKeyById(keySpec);
         if (!key) return p;
-        keys = [key];
       }
       else {
-        keys = getKeyRing().getKeysByUserId(keySpec);
-        if (keys.length === 0) return p;
+        key = getKeyRing().getSecretKeyByEmail(keySpec);
+        if (!key) return p;
       }
 
-      for (let i in keys) {
-        let ot = keys[i].ownerTrust;
-        if (ot !== "u") p.push(keys[i]);
-      }
+      let ot = key.ownerTrust;
+      if (ot !== "u") p.push(key);
 
       return p;
     }, []);
diff --git a/package/windows.jsm b/package/windows.jsm
index 4a49b3a..d1d9ab4 100644
--- a/package/windows.jsm
+++ b/package/windows.jsm
@@ -475,6 +475,10 @@ var EnigmailWindows = {
   openKeyDetails: function(win, keyId, refresh) {
     const keyListObj = {};
 
+    if (!win) {
+      win = this.getBestParentWin();
+    }
+
     keyId = keyId.replace(/^0x/, "");
 
     if (refresh) {
diff --git a/ui/content/enigmailMsgComposeOverlay.js b/ui/content/enigmailMsgComposeOverlay.js
index 2a94bd7..c3d6ef1 100644
--- a/ui/content/enigmailMsgComposeOverlay.js
+++ b/ui/content/enigmailMsgComposeOverlay.js
@@ -4519,7 +4519,7 @@ Enigmail.msg = {
       key = EnigmailKeyRing.getKeyById(this.identity.getCharAttribute("pgpkeyId"));
     }
     else {
-      key = EnigmailKeyRing.getSecretKeyByUserId(this.identity.email);
+      key = EnigmailKeyRing.getSecretKeyByEmail(this.identity.email);
     }
 
     if (key) {
